torch/csrc/api/include/torch/data/dataloader/stateless_impl.h

```diff
@@ -0,0 +1,109 @@
+#pragma once
+
+#include <torch/data/dataloader/impl.h>
+#include <torch/data/worker_exception.h>
+
+#include <torch/csrc/utils/memory.h>
+
+#include <c10/util/Exception.h>
+
+#include <cstddef>
+#include <thread>
+#include <utility>
+
+namespace torch {
+namespace data {
+namespace detail {
+namespace dataloader {
+template <typename Dataset, typename Sampler>
+class StatelessImpl : public Impl<
+                          Dataset,
+                          typename Dataset::BatchType,
+                          typename Sampler::BatchRequestType> {
+ public:
+  using super = Impl<
+      Dataset,
+      typename Dataset::BatchType,
+      typename Sampler::BatchRequestType>;
+  using super::options_;
+  using super::pop_result;
+  using super::prefetch;
+  using super::push_job;
+  using super::workers_;
+  using typename super::BatchRequestType;
+  using typename super::BatchType;
+  using typename super::Result;
+
+  StatelessImpl(Dataset dataset, DataLoaderOptions options, Sampler sampler)
+      : super(std::move(options)), sampler_(std::move(sampler)) {
+    for (size_t w = 0; w < options_.workers; ++w) {
+      // Here we copy the dataset into the worker thread closure. Each worker
+      // has its own copy of the dataset. This means the dataset must be
+      // trivially copiable, or else we don't expect more than one worker to
+      // be in use.
+      workers_.emplace_back(
+          [this, dataset]() mutable { this->worker_thread(dataset); });
+    }
+    if (options_.workers == 0) {
+      main_thread_dataset_ = torch::make_unique<Dataset>(std::move(dataset));
+    }
+  }
+
+  void reset(bool prefetch = true) override {
+    sampler_.reset();
+    super::reset(prefetch);
+  }
+
+ private:
+  optional<BatchType> next() {
+    optional<BatchType> batch;
+    if (options_.workers > 0) {
+      optional<Result> result = pop_result();
+      if (result) {
```

Comment:
Why do we need a while loop in the stateful case, but not in here?

Reply:
Because in the stateless case, the main thread sees that the sampler has exhausted, then notifies the worker threads to die. But in the stateful case, the workers will in non-deterministic order get empty batches once the dataset exhausts. The while loop then has to skip the empty batches in the main thread because some workers may have gotten the last few batches while other workers already got the empty batches, and enqueue those to the main thread first (but we have to process the non-empty batches first of course). The terminating condition in the stateful case is that the result queue will be completely empty, in which the main thread knows to stop as well